### PR TITLE
Develop: api, errors, exceptions, translations

### PR DESCRIPTION
**Pull request by gymad**
## Description

Develop: api, errors, exceptions, translations

## How To Test This

- Switch on developerMode in SuiteCRM (optional step)
- Follow the documentation: https://docs.suitecrm.com/developer/api/version-8/
- Call wrong urls or bean-ids etc. in the request

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

File: lib/API/v8/Controller/ApiController.php [hunk 1]
```diff
@@ -106,126 +108,173 @@ public function __construct(ContainerInterface $containers)
...
-            ->write(json_encode($payload));
     }
 
     /**
      * @param Request $request
      * @param Response $response
      * @param \Exception|ApiException $exception
-     * @return Response
-     * @throws \InvalidArgumentException
+     * @throws RuntimeException
```
Comment (zoltankocsardi) [inline lib/API/v8/Controller/ApiController.php]: missing @return tag

File: lib/API/v8/Controller/OAuth2Controller.php [hunk 2]
```diff
@@ -42,34 +42,41 @@
...
+            $server = $this->containers->get('AuthorizationServer');
+
+            try {
+                return $server->respondToAccessTokenRequest($request, $response);
+            } catch (OAuthServerException $exception) {
+                return $exception->generateHttpResponse($response);
+            }
+            
+        } catch (\Exception $e) {
+            $payload = $this->handleExceptionIntoPayloadError($request, $e, isset($payload) ? $payload : []);
```
Comment (zoltankocsardi) [inline lib/API/v8/Controller/OAuth2Controller.php]: Does this work? The 3rd parameter passed by reference in handleExceptionIntoPayloadError.

File: include/LangText.php [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+     * @param array|null $args
+     * @param boolean|null $log
+     * @param integer $use
+     * @param boolean $throw
+     * @return string
+     */
+    public static function get($key, $args = null, $use = self::USING_ALL_STRINGS, $log = true, $throw = true)
+    {
+        $text = new LangText($key, $args, $use, $log, $throw);
+        $translated = $text->getText();
```
[the remark this answers is not in the document]
Comment (gymad) [reply include/LangText.php]: Yes I could but easiest way to debugging the return value of this function if the return statement get a variable instead a function call.

File: include/LangException.php [hunk 4]
```diff
@@ -38,40 +39,57 @@
...
-        parent::__construct('[IdAlreadyExists] '.$message, $code, $previous);
-    }
+    protected $langMessage;
 
     /**
-     * @return int
+     *
+     * @param string $message
+     * @param integer $code
+     * @param Throwable $previous
```
[the remark this answers is not in the document]
Comment (gymad) [reply include/LangException.php]: Logically a Throwable as it is in the parent core Exception but it's cause a PHP error.

File: include/LangText.php [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    /**
+     *
+     * @global array $app_strings
+     * @global array $mod_strings
+     * @param string|null $key
+     * @param array|null $args
+     * @param integer|null $use
+     * @return string
```
[the remark this answers is not in the document]
Comment (gymad) [reply include/LangText.php]: added

File: include/LangText.php [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    /**
+     *
+     * @param string $key
+     * @param array|null $args
+     * @param boolean|null $log
+     * @param integer $use
+     * @param boolean $throw
+     * @return string
```
[the remark this answers is not in the document]
Comment (gymad) [reply include/LangText.php]: added

File: lib/API/v8/Controller/OAuth2Controller.php (hunk 2 above)
Comment (gymad) [reply lib/API/v8/Controller/OAuth2Controller.php]: & removed

File: include/LangText.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (zoltankocsardi) [reply include/LangText.php]: Fair enough

**Comment by daniel-samson** [comment]
Assessed

**Pull request merged**